src/coreComponents/mesh/CellElementRegion.cpp

```diff
@@ -35,16 +73,163 @@ CellElementRegion::CellElementRegion( string const & name, Group * const parent
 CellElementRegion::~CellElementRegion()
 {}
 
-void CellElementRegion::generateMesh( Group const & cellBlocks )
+
+void CellElementRegion::registerSubRegion( CellBlockABC const & cellBlock )
 {
   Group & elementSubRegions = this->getGroup( viewKeyStruct::elementSubRegions() );
+  // For now, subRegion name must be the same as the cellBlock (so we can match them and reference them in errors).
+  CellElementSubRegion & subRegion =
+    elementSubRegions.registerGroup< CellElementSubRegion >( cellBlock.getName() );
+
+  subRegion.copyFromCellBlock( cellBlock );
+}
+
+string CellElementRegion::getCellBlockAttributeValue( string_view cellBlockName )
+{
+  return string( stringutilities::removeStringAndFollowingContent( cellBlockName, "_" ) );
+}
 
-  for( string const & cellBlockName : this->m_cellBlockNames )
+std::set< string > getAvailableAttributeValues( std::set< string > const & cellBlocksNames )
+{
+  std::set< string > attributeValues;
+  for( auto const & cellBlockName : cellBlocksNames )
   {
-    CellElementSubRegion & subRegion = elementSubRegions.registerGroup< CellElementSubRegion >( cellBlockName );
-    CellBlockABC const & source = cellBlocks.getGroup< CellBlockABC >( subRegion.getName() );
-    subRegion.copyFromCellBlock( source );
+    attributeValues.insert( CellElementRegion::getCellBlockAttributeValue( cellBlockName ) );
   }
+  return attributeValues;
+}
+
+std::set< string >
+CellElementRegion::computeSelectedCellBlocks( std::set< string > const & cellBlocksNames ) const
```

Comment:
You have errors checks at a _lot_ of different places and eventually I really have hard time following the flow of what's going on.
Could you simplify the code into having a `stream/iterable` of `CellBlockABC` that you filter in a generic way using a `predicate` that you'll build from the user input.
I think that will make the code a lot simpler without degrading the use UX.
Otherwise, I think we're good. 😉